gui/test/ruler-tests.cc

```diff
@@ -0,0 +1,370 @@
+
+BOOST_AUTO_TEST_SUITE(Ruler_Tests)
+
+BOOST_AUTO_TEST_CASE(Ruler_creation_horizontal_signal_handlers,
+    * utf::description("Tests that correct signal handlers are registered at creation of a horizontal ruler."))
+{
+    gtk_init(nullptr, nullptr);
+    // Register a new ruler with a dummy drawing area
+    GtkWidget* drawingArea = gtk_drawing_area_new ();
+    Ruler::Ptr ruler = Ruler::create(Ruler::HORIZONTAL, drawingArea);
+    // Check that the appropriate signals are connected
+
+    // Currently, this test case only checks that *a* signal handler is connected
+    // for both the "draw" signal and "size-allocate" signal, with a pointer
+    // to the ruler as data.
+    // It does not check whether drawCallback is connected to "draw" and
+    // sizeAllocateCallback to "size-allocate". It probably should, but I can't figure
+    // out these weird types.
+    auto mask = static_cast<GSignalMatchType>(G_SIGNAL_MATCH_ID | G_SIGNAL_MATCH_DATA);
+    guint drawID = g_signal_lookup("draw", GTK_TYPE_DRAWING_AREA);
+    guint sizeAllocateID = g_signal_lookup("size-allocate", GTK_TYPE_DRAWING_AREA);
+    // Check that a signal handler is connected for the "draw" signal, with a pointer to ruler as data
+    BOOST_CHECK(g_signal_handler_find(drawingArea, mask, drawID, 0, nullptr, nullptr, ruler.get()) != 0);
+    // Check that a signal handler is connected for the "size-allocate" signal, with a pointer to ruler as data
+    BOOST_CHECK(g_signal_handler_find(drawingArea, mask, sizeAllocateID, 0, nullptr, nullptr, ruler.get()) != 0);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_creation_vertical_signal_handlers,
+     * utf::description("Tests that correct signal handlers are registered at creation of a vertical ruler."))
+{
+  gtk_init(nullptr, nullptr);
+  // Register a new ruler with a dummy drawing area
+  GtkWidget* drawingArea = gtk_drawing_area_new ();
+  Ruler::Ptr ruler = Ruler::create(Ruler::VERTICAL, drawingArea);
+  // Check that the appropriate signals are connected
+  auto mask = static_cast<GSignalMatchType>(G_SIGNAL_MATCH_ID | G_SIGNAL_MATCH_DATA);
+  guint drawID = g_signal_lookup("draw", GTK_TYPE_DRAWING_AREA);
+  guint sizeAllocateID = g_signal_lookup("size-allocate", GTK_TYPE_DRAWING_AREA);
+  // Check that a signal handler is connected for the "draw" signal, with a pointer to ruler as data
+  BOOST_CHECK(g_signal_handler_find(drawingArea, mask, drawID, 0, nullptr, nullptr, ruler.get()) != 0);
+  // Check that a signal handler is connected for the "size-allocate" signal, with a pointer to ruler as data
+  BOOST_CHECK(g_signal_handler_find(drawingArea, mask, sizeAllocateID, 0, nullptr, nullptr, ruler.get()) != 0);
+}
+
+///////////////
+// Testing an all-positive range
+
+BOOST_AUTO_TEST_CASE(Ruler_intervalCalculation_0_to_10_width_1920px,
+     * utf::description("Tests the interval for a range [0, 10] for a ruler of width 1920px is 1"))
+{
+    BOOST_CHECK(RulerCalculations::calculateInterval(0, 10, 1920) == 1);
```

Comment:
All of these tests are good, in the sense that they test the functionality of the ruler. However, for someone new to the code, they are a bit hard to read.

When getting to know new code, some people prefer to read the tests, because the tests tell them how the code should be used, what they can expect from the code. Therefore, ideally, a test should refer a requirement. For this test, possibly something like "When the ruler is large, compared to the range, we use an interval of 1". When at all possible, use the requirement as the name of the test.

All tests here test a combination of (4) numbers, without explaining why those numbers are correct. In that sense they are not contributing to the understanding of the reader, and may also be a bit hard to maintain. When the requirements change, which tests should be changed?

This can be improved by not hardcoding the numberer, but by computing them. For example, given
```
int startRange = 0;
int endRange = 10;
```
you might do `int largeRulerWidth = (endRange-startRange)*MIN_TICK_DISTANCE`, and then check that the computed interval is 1. Then, from reading the code, we get a sense of why these numbers are the way they are.

Lastly, there is a lot of duplication in these tests. The numbers that are being tested are repeated in both the test name, the test description and the test implementation. I think for most of these testcases you could omit the description without reducing understandability.